lib/Conversion/GPUToROCM/ConvertKernelFuncToHSACO.cpp

```diff
@@ -0,0 +1,409 @@
+                               llvm::TargetMachine &targetMachine);
+
+  OwnedHSACO emitModuleToHSACOForTesting(llvm::Module &llvmModule,
+                                         FuncOp &function);
+
+  rocm::HSACOGeneratorConfig config;
+};
+
+// get the "-mcpu" option string corresponding to the given AMDGPU version enum
+std::string GetMcpuOptionString(rocm::AMDGPUVersion v) {
+  switch (v) {
+  case rocm::AMDGPUVersion::GFX900:
+    return "gfx900";
+    break;
+  }
+  return "<invalid AMDGPU version>";
+}
+
+// get filename for file containing the AMDGPU version specific bitcodes
+std::string GetBitcodeFilename(rocm::AMDGPUVersion v) {
+  switch (v) {
+  case rocm::AMDGPUVersion::GFX900:
+    return "oclc_isa_version_900.amdgcn.bc";
+    break;
+  }
+  return "<invalid AMDGPU version>";
+}
+
+// get the option string corresponding to the given HSACO version enum
+std::string GetCodeObjectOptionString(rocm::HSACOVersion v) {
+  switch (v) {
+  case rocm::HSACOVersion::V3:
+    return "-code-object-v3";
+    break;
+  }
+  return "invalid HSACO version";
+}
+
+// Gets the ROCm-Device-Libs filenames for a particular AMDGPU version.
+static std::vector<std::string>
+GetROCDLPaths(const rocm::AMDGPUVersion amdgpuVersion,
+              const std::string &rocdlDir) {
+
+  // AMDGPU version-neutral bitcodes.
+  static std::vector<std::string> *rocdlFilenames =
+      new std::vector<std::string>({"hc.amdgcn.bc", "opencl.amdgcn.bc",
+                                    "ocml.amdgcn.bc", "ockl.amdgcn.bc",
+                                    "oclc_finite_only_off.amdgcn.bc",
+                                    "oclc_daz_opt_off.amdgcn.bc",
+                                    "oclc_correctly_rounded_sqrt_on.amdgcn.bc",
+                                    "oclc_unsafe_math_off.amdgcn.bc",
+                                    "oclc_wavefrontsize64_on.amdgcn.bc"});
+
+  // Construct full path to ROCDL bitcode libraries.
+  std::vector<std::string> result;
+  for (auto &filename : *rocdlFilenames) {
+    llvm::SmallString<128> appendedPath;
+    llvm::sys::path::append(appendedPath, rocdlDir, filename);
+    result.push_back(appendedPath.c_str());
+  }
+
+  // Add AMDGPU version-specific bitcodes.
+  llvm::SmallString<128> appendedPath;
+  llvm::sys::path::append(appendedPath, rocdlDir,
+                          GetBitcodeFilename(amdgpuVersion));
+  result.push_back(appendedPath.c_str());
+
+  return std::move(result);
+}
+
+std::unique_ptr<llvm::Module>
+LoadBitcodeModule(const std::string &filename, llvm::LLVMContext &llvmContext) {
+
+  llvm::SMDiagnostic diagnostic;
+  std::unique_ptr<llvm::Module> bitcodeModule(
+      llvm::parseIRFile(llvm::StringRef(filename.data(), filename.size()),
+                        diagnostic, llvmContext));
+
+  if (bitcodeModule == nullptr) {
+    llvm::errs() << diagnostic.getFilename().str() << ":"
+                 << diagnostic.getLineNo() << ":" << diagnostic.getColumnNo()
```

Comment:
Can we reencode these with a proper MLIR location and emit them as diagnostics? Possibly as a note attached to a diagnostic for the current module being converted.

Reply:
Can we leave this as is for the time being?
This part of the code is expected to be in flux, and I need to read up on how to properly MLIR diagnostics. We can fix this in a follow up PR (there are a few other places that will require a similar change).

thanks